lib/MusicBrainz/Server/Data/Role/EntityCache.pm

```diff
@@ -1,38 +1,131 @@
 
-role {
+sub _id_cache_info {
+    my ($self, $prop) = @_;
 
-    my $params = shift;
+    my $type = $self->_type;
+    if ($type && (my $entity_properties = $ENTITIES{$type})) {
+        if (exists $entity_properties->{cache}) {
+            return $entity_properties->{cache}{$prop};
+        }
+    }
+}
 
-    with 'MusicBrainz::Server::Data::Role::EntityCacheBase';
+has _id_cache_id => (
+    is => 'ro',
+    isa => 'Maybe[Str]',
+    lazy => 1,
+    default => sub { shift->_id_cache_info('id') },
+);
 
-    method '_id_cache_prefix' => sub { $params->{prefix} };
+has _id_cache_prefix => (
+    is => 'ro',
+    isa => 'Maybe[Str]',
+    lazy => 1,
+    default => sub { shift->_id_cache_info('prefix') },
+);
 
-    method '_add_to_cache' => sub
-    {
-        my ($self, $cache, %data) = @_;
-        my @tmp;
-        foreach my $id (keys %data) {
-            my $key = $self->_id_cache_prefix . ':' . $id;
-            push @tmp, [$key, $data{$id}];
+around get_by_ids => sub {
+    my ($orig, $self, @ids) = @_;
+    return {} unless grep { defined && $_ } @ids;
+    my %ids = map { $_ => 1 } @ids;
+    my @keys = map { $self->_id_cache_prefix . ':' . $_ } keys %ids;
+    my $cache = $self->c->cache($self->_id_cache_prefix);
+    my %data = %{$cache->get_multi(@keys)};
+    my %result;
+    foreach my $key (keys %data) {
+        my @key = split /:/, $key;
+        my $id = $key[1];
+        $result{$id} = $data{$key};
+        delete $ids{$id};
+    }
+    if (%ids) {
+        my $data = $self->$orig(keys %ids) || {};
+        foreach my $id (keys %$data) {
+            $result{$id} = $data->{$id};
         }
-        $cache->set_multi(@tmp);
-    };
+        $self->_add_to_cache($cache, %$data);
+    }
+    return \%result;
+};
 
+after update => sub {
+    my ($self, $id) = @_;
+    $self->_delete_from_cache($id);
 };
 
+after delete => sub {
+    my ($self, @ids) = @_;
+    $self->_delete_from_cache(@ids);
+};
+
+after merge => sub {
+    my ($self, @ids) = @_;
+    $self->_delete_from_cache(@ids);
+};
+
+sub _create_cache_entries {
+    my ($self, $data) = @_;
+
+    my $cache_id = $self->_id_cache_id;
+    my $cache_prefix = $self->_id_cache_prefix . ':';
+    my @entries;
+    for my $id (keys %{$data}) {
+        # MBS-7241
+        my $got_lock = $self->c->sql->select_single_value(
+            'SELECT pg_try_advisory_xact_lock(?, ?)',
+            $cache_id,
+            $id,
+        );
+        if ($got_lock) {
+            push @entries, [$cache_prefix . $id, $data->{$id}];
+        }
+    }
+    @entries;
+}
+
+sub _add_to_cache {
+    my ($self, $cache, %data) = @_;
+
+    $cache->set_multi($self->_create_cache_entries(\%data));
+}
+
+sub _delete_from_cache {
+    my ($self, @ids) = @_;
+
+    @ids = uniq grep { defined } @ids;
+    return unless @ids;
+
+    my $cache_id = $self->_id_cache_id;
+    my $cache_prefix = $self->_id_cache_prefix . ':';
+    my @keys;
+
+    for my $id (@ids) {
+        if (looks_like_number($id)) {
```

Comment:
Since gids are passed to this method too. I'm pretty sure it's safe that we only lock things based on the row id, since if the gid isn't cached, it has to call `get_by_id` which acquires a lock on the id.